File: src/lib/protocol/event.ts
export type NostrTag = readonly string[];

export type UnsignedNostrEvent = {
  readonly pubkey: string;
  readonly created_at: number;
  readonly kind: number;
  readonly tags: readonly NostrTag[];
  readonly content: string;
};

export type NostrEvent = UnsignedNostrEvent & {
  readonly id: string;
  readonly sig: string;
};

export type EventFramePolicy = {
  readonly maxEventContentBytes: number;
  readonly maxEventTags: number;
  readonly maxTagFields: number;
  readonly maxTagFieldBytes: number;
};

export type EventValidation =
  | { ok: true; event: NostrEvent }
  | {
      ok: false;
      code: 'not_object' | 'bad_field' | 'bad_tag';
      message: string;
    };

const hex64 = /^[0-9a-f]{64}$/;
const hex128 = /^[0-9a-f]{128}$/;

export function parseNostrEvent(
  value: unknown,
  policy?: EventFramePolicy,
): EventValidation {
  if (!isRecord(value)) return fail('not_object', 'event must be an object');
  const base = parseUnsignedEvent(value, policy);
  if (!base.ok) return base;
  if (!isHex(value.id, hex64))
    return fail('bad_field', 'id must be 32-byte lowercase hex');
  if (!isHex(value.sig, hex128))
    return fail('bad_field', 'sig must be 64-byte lowercase hex');
  return { ok: true, event: { ...base.event, id: value.id, sig: value.sig } };
}

export function parseUnsignedEvent(
  value: unknown,
  policy?: EventFramePolicy,
):
  | { ok: true; event: UnsignedNostrEvent }
  | {
      ok: false;
      code: 'not_object' | 'bad_field' | 'bad_tag';
      message: string;
    } {
  if (!isRecord(value)) return fail('not_object', 'event must be an object');
  if (!isHex(value.pubkey, hex64))
    return fail('bad_field', 'pubkey must be 32-byte lowercase hex');
  if (!isUnixTime(value.created_at))
    return fail('bad_field', 'created_at must be a non-negative integer');
  if (!isKind(value.kind))
    return fail('bad_field', 'kind must be a non-negative integer');
  if (!Array.isArray(value.tags))
    return fail('bad_field', 'tags must be an array');
  if (typeof value.content !== 'string')
    return fail('bad_field', 'content must be a string');
  if (
    policy &&
    !utf8ByteLengthWithin(value.content, policy.maxEventContentBytes).within
  )
    return fail(
      'bad_field',
      `content exceeds ${policy.maxEventContentBytes} bytes`,
    );
  const tags = parseTags(value.tags, policy);
  if (!tags.ok) return tags;
  return {
    ok: true,
    event: {
      pubkey: value.pubkey,
      created_at: value.created_at,
      kind: value.kind,
      tags: tags.tags,
      content: value.content,
    },
  };
}

export function compareEventsDesc(a: NostrEvent, b: NostrEvent): number {
  if (a.created_at !== b.created_at) return b.created_at - a.created_at;
  return a.id.localeCompare(b.id);
}

export function isEventId(value: string): boolean {
  return hex64.test(value);
}

export function isPubkey(value: string): boolean {
  return hex64.test(value);
}

function parseTags(
  value: unknown[],
  policy?: EventFramePolicy,
):
  | { ok: true; tags: readonly NostrTag[] }
  | { ok: false; code: 'bad_tag'; message: string } {
  if (policy && value.length > policy.maxEventTags)
    return {
      ok: false,
      code: 'bad_tag',
      message: `tags exceed ${policy.maxEventTags} entries`,
    };
  const tags: string[][] = [];
  for (const tag of value) {
    if (!Array.isArray(tag))
      return { ok: false, code: 'bad_tag', message: 'tag must be an array' };
    if (policy && tag.length > policy.maxTagFields)
      return {
        ok: false,
        code: 'bad_tag',
        message: `tag exceeds ${policy.maxTagFields} fields`,
      };
    if (!tag.every((part) => typeof part === 'string')) {
      return {
        ok: false,
        code: 'bad_tag',
        message: 'tag entries must be strings',
      };
    }
    if (
      policy &&
      tag.some(
        (part) => !utf8ByteLengthWithin(part, policy.maxTagFieldBytes).within,
      )
    )
      return {
        ok: false,
        code: 'bad_tag',
        message: `tag field exceeds ${policy.maxTagFieldBytes} bytes`,
      };
    tags.push([...tag]);
  }
  return { ok: true, tags };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isHex(value: unknown, pattern: RegExp): value is string {
  return typeof value === 'string' && pattern.test(value);
}

function isUnixTime(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 0;
}

function isKind(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 0;
}

function fail(
  code: 'not_object' | 'bad_field',
  message: string,
): EventValidation {
  return { ok: false, code, message };
}

function utf8ByteLengthWithin(
  text: string,
  limit: number,
): {
  readonly within: boolean;
} {
  let bytes = 0;
  for (let index = 0; index < text.length; index++) {
    const code = text.charCodeAt(index);
    if (code <= 0x7f) bytes += 1;
    else if (code <= 0x7ff) bytes += 2;
    else if (
      code >= 0xd800 &&
      code <= 0xdbff &&
      isLowSurrogate(text.charCodeAt(index + 1))
    ) {
      bytes += 4;
      index++;
    } else {
      bytes += 3;
    }
    if (bytes > limit) return { within: false };
  }
  return { within: true };
}

function isLowSurrogate(code: number): boolean {
  return code >= 0xdc00 && code <= 0xdfff;
}

File: crates/lkjstr-protocol/src/event.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{event_tags::parse_tags, is_lower_hex};

pub type NostrTag = Vec<String>;

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct UnsignedNostrEvent {
    pub pubkey: String,
    pub created_at: u64,
    pub kind: u64,
    pub tags: Vec<NostrTag>,
    pub content: String,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct NostrEvent {
    pub id: String,
    pub pubkey: String,
    pub created_at: u64,
    pub kind: u64,
    pub tags: Vec<NostrTag>,
    pub content: String,
    pub sig: String,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct EventFramePolicy {
    pub max_event_content_bytes: usize,
    pub max_event_tags: usize,
    pub max_tag_fields: usize,
    pub max_tag_field_bytes: usize,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum EventValidationCode {
    NotObject,
    BadField,
    BadTag,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct EventValidationError {
    pub code: EventValidationCode,
    pub message: String,
}

pub fn parse_nostr_event_json(
    json: &str,
    policy: Option<&EventFramePolicy>,
) -> Result<NostrEvent, EventValidationError> {
    let value = serde_json::from_str(json).map_err(|error| EventValidationError {
        code: EventValidationCode::NotObject,
        message: format!("event json is invalid: {error}"),
    })?;
    parse_nostr_event_value(&value, policy)
}

pub fn parse_nostr_event_value(
    value: &Value,
    policy: Option<&EventFramePolicy>,
) -> Result<NostrEvent, EventValidationError> {
    let base = parse_unsigned_event_value(value, policy)?;
    let id = string_field(value, "id", "id must be 32-byte lowercase hex")?;
    if !is_event_id(&id) {
        return bad_field("id must be 32-byte lowercase hex");
    }
    let sig = string_field(value, "sig", "sig must be 64-byte lowercase hex")?;
    if !is_signature(&sig) {
        return bad_field("sig must be 64-byte lowercase hex");
    }
    Ok(NostrEvent {
        id,
        pubkey: base.pubkey,
        created_at: base.created_at,
        kind: base.kind,
        tags: base.tags,
        content: base.content,
        sig,
    })
}

pub fn parse_unsigned_event_value(
    value: &Value,
    policy: Option<&EventFramePolicy>,
) -> Result<UnsignedNostrEvent, EventValidationError> {
    if !value.is_object() {
        return fail(EventValidationCode::NotObject, "event must be an object");
    }
    let pubkey = string_field(value, "pubkey", "pubkey must be 32-byte lowercase hex")?;
    if !is_pubkey(&pubkey) {
        return bad_field("pubkey must be 32-byte lowercase hex");
    }
    let created_at = u64_field(
        value,
        "created_at",
        "created_at must be a non-negative integer",
    )?;
    let kind = u64_field(value, "kind", "kind must be a non-negative integer")?;
    let content = string_field(value, "content", "content must be a string")?;
    if let Some(frame_policy) = policy
        && content.len() > frame_policy.max_event_content_bytes
    {
        return bad_field(format!(
            "content exceeds {} bytes",
            frame_policy.max_event_content_bytes
        ));
    }
    let tags = parse_tags(value.get("tags"), policy)?;
    Ok(UnsignedNostrEvent {
        pubkey,
        created_at,
        kind,
        tags,
        content,
    })
}

pub fn compare_events_desc(a: &NostrEvent, b: &NostrEvent) -> std::cmp::Ordering {
    b.created_at
        .cmp(&a.created_at)
        .then_with(|| a.id.cmp(&b.id))
}

pub fn is_event_id(value: &str) -> bool {
    is_fixed_lower_hex(value, 64)
}

pub fn is_pubkey(value: &str) -> bool {
    is_fixed_lower_hex(value, 64)
}

pub fn is_signature(value: &str) -> bool {
    is_fixed_lower_hex(value, 128)
}

fn string_field(
    value: &Value,
    field: &str,
    message: &'static str,
) -> Result<String, EventValidationError> {
    value
        .get(field)
        .and_then(Value::as_str)
        .map(ToOwned::to_owned)
        .ok_or_else(|| EventValidationError {
            code: EventValidationCode::BadField,
            message: message.to_owned(),
        })
}

fn u64_field(
    value: &Value,
    field: &str,
    message: &'static str,
) -> Result<u64, EventValidationError> {
    value
        .get(field)
        .and_then(Value::as_u64)
        .ok_or_else(|| EventValidationError {
            code: EventValidationCode::BadField,
            message: message.to_owned(),
        })
}

fn is_fixed_lower_hex(value: &str, len: usize) -> bool {
    value.len() == len && is_lower_hex(value)
}

pub(crate) fn bad_field<T>(message: impl Into<String>) -> Result<T, EventValidationError> {
    fail(EventValidationCode::BadField, message)
}

pub(crate) fn bad_tag<T>(message: impl Into<String>) -> Result<T, EventValidationError> {
    fail(EventValidationCode::BadTag, message)
}

fn fail<T>(
    code: EventValidationCode,
    message: impl Into<String>,
) -> Result<T, EventValidationError> {
    Err(EventValidationError {
        code,
        message: message.into(),
    })
}
